### PR TITLE
ENG-10945-count-anywhere-mviews

### PR DESCRIPTION
**Pull request by dm89**
enable count(*) to be placed anywhere in materialized views

### CODE REVIEW

File: src/catgen/in/javasrc/CatalogDiffEngine.java [hunk 1]
```diff
@@ -1040,6 +1040,10 @@ else if (field.equals("type") || field.equals("size") || field.equals("inbytes")
         }
 
         else if (suspect instanceof MaterializedViewInfo) {
+            // ENG-10945 Allow materialized views to change the count star position they use dynamically.
+            if( field.equals("countStarColumnIndex") )
```
Comment (yzhang1991) [inline src/catgen/in/javasrc/CatalogDiffEngine.java]: Formatting:
```java
if (field.equals("countStarColumnIndex")) {
    return null;
}
```

File: src/ee/storage/MaterializedViewHandler.cpp [hunk 2]
```diff
@@ -294,15 +298,20 @@ namespace voltdb {
...
-        m_updatedTuple.setNValue(m_groupByColumnCount, newCount);
-        int aggOffset = m_groupByColumnCount + 1;
+        m_updatedTuple.setNValue(m_countStarColumnIndex, newCount);
+
+        int aggOffset = m_groupByColumnCount;
         NValue newValue;
         if (newCount.isZero()) {
             // no group by key, no rows, aggs will be null except for count().
             for (int aggIndex = 0, columnIndex = aggOffset; aggIndex < m_aggColumnCount; aggIndex++, columnIndex++) {
+                if (m_aggTypes[aggIndex] == EXPRESSION_TYPE_AGGREGATE_COUNT_STAR) {
```
Comment (yzhang1991) [inline src/ee/storage/MaterializedViewHandler.cpp]: What if there are multiple COUNT(\*) in the definition? In your code, only the first COUNT(*) column will have the correct value.

File: src/ee/storage/MaterializedViewHandler.cpp [hunk 3]
```diff
@@ -323,6 +332,8 @@ namespace voltdb {
 
                 if (! deltaValue.isNull()) {
                     switch(aggType) {
+                        case EXPRESSION_TYPE_AGGREGATE_COUNT_STAR: // already done above
```
Comment (yzhang1991) [inline src/ee/storage/MaterializedViewHandler.cpp]: Again, with your changes, I can define a view with multiple COUNT(*) columns, and only the first COUNT(*) will have the correct value.

File: src/ee/storage/MaterializedViewHandler.cpp [hunk 4]
```diff
@@ -397,8 +408,10 @@ namespace voltdb {
                                     " looking for a tuple in the view and"
                                     " expected to find it but didn't", name.c_str());
             }
-            NValue existingCount = m_existingTuple.getNValue(m_groupByColumnCount);
-            NValue deltaCount = deltaTuple.getNValue(m_groupByColumnCount);
+            NValue existingCount = m_existingTuple.getNValue(m_countStarColumnIndex);
+            NValue deltaCount = deltaTuple.getNValue(m_countStarColumnIndex);
+
+            // not sure why we have m_groupByColumnCount > 0 - should it be m_countStarColumnIndex?
```
Comment (yzhang1991) [inline src/ee/storage/MaterializedViewHandler.cpp]: If there is no GROUP-BY column, the aggregation result will always have one row, even if the count = 0.
Please remove this line of the comment.

File: src/ee/storage/MaterializedViewTriggerForInsert.cpp [hunk 5]
```diff
@@ -89,9 +90,10 @@ void MaterializedViewTriggerForInsert::updateDefinition(PersistentTable *destTab
 }
 
 NValue MaterializedViewTriggerForInsert::getAggInputFromSrcTuple(int aggIndex,
+                                                                 int aggExprOffset,
```
Comment (yzhang1991) [inline src/ee/storage/MaterializedViewTriggerForInsert.cpp]: Is it necessary to have this additional parameter?
You are treating the COUNT(\*) column very differently, in many places if the user defines the view with multiple COUNT(\*) you will have problems.

File: src/frontend/org/voltdb/compiler/MaterializedViewProcessor.java [hunk 6]
```diff
@@ -504,29 +514,36 @@ private void checkViewMeetsSpec(String viewName, ParsedSelectStmt stmt) throws V
...
         // calls.  But these are not any aggregate calls. They
         // must be count(), min(), max() or sum().
-        for (i++; i < displayColCount; i++) {
+        for (; i < displayColCount; i++) {
             ParsedColInfo outcol = stmt.m_displayColumns.get(i);
             // Note that this expression does not catch all aggregates.
-            // An instance of count(*) here, or avg() would cause the
-            // exception.  We just required count(*) above, but a
+            // An instance of avg() would cause the exception.
+            // ENG-10945 - We can have count(*) anywhere after the group by columns, but a
```
Comment (dm89) [inline src/frontend/org/voltdb/compiler/MaterializedViewProcessor.java]: @yzhang1991  Here is the guard for only one count(*) right now

File: tests/frontend/org/voltdb/regressionsuites/TestGroupByComplexMaterializedViewSuite.java [hunk 7]
```diff
@@ -1598,7 +1598,7 @@ public TestGroupByComplexMaterializedViewSuite(String name) {
                 "PRIMARY KEY (F_PKEY) ); " +
 
                 "CREATE VIEW V1 (V_D1_PKEY, V_D2_PKEY, V_D3_PKEY, V_F_PKEY, CNT, SUM_V1, SUM_V2, SUM_V3) " +
-                "AS SELECT F_D1, F_D2, F_D3, F_PKEY, COUNT(*) + 1, SUM(F_VAL1), SUM(F_VAL2), SUM(F_VAL3) " +
```
Comment (yzhang1991) [inline tests/frontend/org/voltdb/regressionsuites/TestGroupByComplexMaterializedViewSuite.java]: Change the column name above, `CNT`

Comment (yzhang1991) [inline tests/frontend/org/voltdb/regressionsuites/TestGroupByComplexMaterializedViewSuite.java]: Why you change this case in this way? Do you have COUNT(*) + 1 case covered anywhere else?

Comment (dm89) [inline tests/frontend/org/voltdb/regressionsuites/TestGroupByComplexMaterializedViewSuite.java]: I changed it for getting different error messages. I added another view now and kept COUNT(*) + 1

**Pull request merged**